### PR TITLE
NMS-8142: Fix collapsible criteria is not persisted

### PR DESCRIPTION
**Pull request by pdzaccone**
* JIRA: https://issues.opennms.org/browse/NMS-8142

### CODE REVIEW

File: features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/support/SavedHistory.java [hunk 1]
```diff
@@ -201,6 +230,20 @@ public void apply(GraphContainer graphContainer, Collection<HistoryOperation> op
                 LOG.warn("Failed to perform applyHistory() operation", e);
             }
         }
+
+        List<SearchProvider> searchProviders = serviceLocator.findServices(SearchProvider.class,null);
+        for (SearchProvider searchProvider : searchProviders) {
+            if (!(searchProvider instanceof HistoryAwareSearchProvider))
+                continue;
```
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/support/SavedHistory.java]: Can we not use `continue` statements and invert the if-statement and include the for block below, to make it more clear what is going on.

File: features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/support/SavedHistory.java [hunk 2]
```diff
@@ -201,6 +230,20 @@ public void apply(GraphContainer graphContainer, Collection<HistoryOperation> op
...
+            for (SearchResult searchQuery : m_searchQueries) {
+                if (searchProvider.getSearchProviderNamespace().equals(searchQuery.getNamespace()) || searchProvider.contributesTo(searchQuery.getNamespace())) {
+                    Criteria searchCriteria = ((HistoryAwareSearchProvider)searchProvider).getCriteriaFromQuery(searchQuery);
+                    if (searchCriteria != null)
```
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/support/SavedHistory.java]: Please always use brackets (`{´ and `}`) for each if.

File: features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/topo/SearchCriteria.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+ * For more information contact:
+ *     OpenNMS(R) Licensing <license@opennms.org>
+ *     http://www.opennms.org/
+ *     http://www.opennms.com/
+ *******************************************************************************/
+
+package org.opennms.features.topology.api.topo;
+
+public interface SearchCriteria
+{
```
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/topo/SearchCriteria.java]: Please add the brackets at the end of the class/interface definition line.

File: features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/topo/SearchResult.java [hunk 4]
```diff
@@ -53,7 +70,11 @@ public SearchResult(SearchResult result) {
         setCollapsible(result.isCollapsible());
     }
 
-    public final String getId() {
+	public SearchResult(SearchCriteria c) {
```
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/topo/SearchResult.java]: Please use a more meaningful name, like `criteria` or `searchCriteria`.

File: features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/AlarmProviderDao.java [hunk 5]
```diff
@@ -26,29 +26,23 @@
...
- * @author <a href=mailto:seth@opennms.org>Seth Leger</a>
- *
- */
-public class CategoryHopCriteriaFactory {
+import org.opennms.netmgt.dao.api.AlarmDao;
+import org.opennms.netmgt.model.OnmsAlarm;
 
-	private final CategoryDao m_categoryDao;
-	private final NodeDao m_nodeDao;
+public class AlarmProviderDao implements AlarmProvider {
```
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/AlarmProviderDao.java]: We should name it `AlarmDaoProvider`.

File: features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/AlarmSearchProvider.java [hunk 6]
```diff
@@ -86,8 +85,16 @@ public String getSearchProviderNamespace() {
     public boolean contributesTo(String namespace) {
         return "nodes".equals(namespace);
     }
-    
-    public class AlarmSearchResult extends SearchResult {
+
+	@Override
+	public Criteria getCriteriaFromQuery(SearchResult input) {
+		// TODO RS Verify that this actually works... node label seems to be missing
```
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/AlarmSearchProvider.java]: Please remove TODO

File: features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/AlarmSearchProvider.java [hunk 7]
```diff
@@ -86,8 +85,16 @@ public String getSearchProviderNamespace() {
...
+		AlarmHopCriteria criteria = createCriteria(input);
```
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/AlarmSearchProvider.java]: the `createCriteria` method should set the id of the criteria based on the input's id. Then the next line can be removed.

File: features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/AlarmSearchProvider.java [hunk 8]
```diff
@@ -352,13 +359,16 @@ private AlarmHopCriteria findCriterion(SearchResult result, GraphContainer conta
...
+                return criterion;
             }
         }
         return null;
     }
-	
-	
+
+    private AlarmHopCriteria createCriteria(SearchResult searchResult)
+	{
```
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/AlarmSearchProvider.java]: Bracket location should be at the end of the line before.

File: features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/CategorySearchProvider.java [hunk 9]
```diff
@@ -113,31 +108,30 @@ public boolean supportsPrefix(String searchPrefix) {
 
     @Override
     public void addVertexHopCriteria(SearchResult searchResult, GraphContainer container) {
-        CategoryHopCriteria criteria = m_categoryHopFactory.getCriteria(searchResult.getLabel());
+        CategoryHopCriteria criteria = createCriteria(searchResult.getLabel());
```
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/CategorySearchProvider.java]: Can the `createCriteria` method use `SearchResult` instead of `searchResult.getLabel()` instead?
The method could then also set the id.

File: features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/CategorySearchProvider.java [hunk 10]
```diff
@@ -113,31 +108,30 @@ public boolean supportsPrefix(String searchPrefix) {
...
-    }
-
     public void setHiddenCategoryPrefix(String prefix) {
         m_hiddenCategoryPrefix = prefix;
     }
 
     @Override
+    public Criteria getCriteriaFromQuery(SearchResult input) {
+        CategoryHopCriteria c = createCriteria(input.getLabel());
+        c.setId(input.getId());
```
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/CategorySearchProvider.java]: same as above.

File: features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/CategoryVertexProvider.java [hunk 11]
```diff
@@ -26,37 +26,16 @@
...
-	 * 
-	 * FIXME these should be static operations.
-	 * 
-	 * @param alarmQuery
-	 * @return
-	 */
-	public AlarmHopCriteria createCriteria(AlarmSearchResult result) {
-		return new AlarmHopCriteria(result, m_alarmDao);
-	}
+public interface CategoryVertexProvider {
```
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/CategoryVertexProvider.java]: This should be named `CategoryProvider`

File: features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/CategoryVertexProviderDao.java [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+import java.util.Collection;
+import java.util.List;
+import java.util.Objects;
+
+import org.opennms.netmgt.dao.api.CategoryDao;
+import org.opennms.netmgt.dao.api.NodeDao;
+import org.opennms.netmgt.model.OnmsCategory;
+import org.opennms.netmgt.model.OnmsNode;
+
+public class CategoryVertexProviderDao implements CategoryVertexProvider {
```
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/CategoryVertexProviderDao.java]: This should be named `CategoryDaoProvider`.

File: features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/CategoryVertexProviderDao.java [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+    private final CategoryDao categoryDao;
+    private final NodeDao nodeDao;
+
+    public CategoryVertexProviderDao() {
```
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/CategoryVertexProviderDao.java]: We should not have a default constructor.

File: features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/IpLikeProvider.java [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+ *******************************************************************************/
+
+package org.opennms.features.topology.app.internal;
+
+import java.util.List;
+
+import org.opennms.core.criteria.Criteria;
+import org.opennms.netmgt.model.OnmsIpInterface;
+
+public interface IpLikeProvider {
```
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/IpLikeProvider.java]: This should be called `IpInterfaceProvider`.

File: features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/IpLikeProviderDao.java [hunk 15]
```diff
@@ -26,37 +26,24 @@
...
- * @author <a href=mailto:david@opennms.org>David Hustace</a>
- *
- */
-public class IpLikeHopCriteriaFactory {
+import org.opennms.core.criteria.Criteria;
+import org.opennms.netmgt.dao.api.IpInterfaceDao;
+import org.opennms.netmgt.model.OnmsIpInterface;
 
-	private final IpInterfaceDao m_ipInterfaceDao;
+public class IpLikeProviderDao implements IpLikeProvider {
```
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/IpLikeProviderDao.java]: This should be called `IpInterfaceDaoProvider`

File: features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/IpLikeSearchProvider.java [hunk 16]
```diff
@@ -252,8 +250,8 @@ public void onDefocusSearchResult(SearchResult searchResult, OperationContext op
 	@Override
 	public void addVertexHopCriteria(SearchResult searchResult, GraphContainer container) {
     	LOG.debug("SearchProvider->addVertexHopCriteria: called with search result: '{}'", searchResult);
-		
-		IpLikeHopCriteria criterion = m_ipLikeHopFactory.createCriteria(searchResult.getLabel());
+
+		IpLikeHopCriteria criterion = createCriteria(searchResult.getLabel());
```
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/IpLikeSearchProvider.java]: Same as above. `createCriteria(String)` vs `createCriteria(SearchResult)`

File: features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/IpLikeSearchProvider.java [hunk 17]
```diff
@@ -296,6 +294,13 @@ public void onToggleCollapse(SearchResult searchResult, GraphContainer graphCont
         
 	}
 
+	@Override
+	public org.opennms.features.topology.api.topo.Criteria getCriteriaFromQuery(SearchResult input) {
+		IpLikeHopCriteria c = createCriteria(input.getQuery());
```
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/IpLikeSearchProvider.java]: see above.

**Review by mvrueden — CHANGES_REQUESTED** [review]

File: features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/support/SavedHistory.java [hunk 18]
```diff
@@ -201,6 +230,21 @@ public void apply(GraphContainer graphContainer, Collection<HistoryOperation> op
...
+
+        List<SearchProvider> searchProviders = serviceLocator.findServices(SearchProvider.class,null);
+        for (SearchProvider searchProvider : searchProviders) {
+            if (searchProvider instanceof HistoryAwareSearchProvider) {
+                for (SearchResult searchQuery : m_searchQueries) {
+                    if (searchProvider.getSearchProviderNamespace().equals(searchQuery.getNamespace()) || searchProvider.contributesTo(searchQuery.getNamespace())) {
+                        Criteria searchCriteria = ((HistoryAwareSearchProvider)searchProvider).getCriteriaFromQuery(searchQuery);
+                        if (searchCriteria != null) {
+                            graphContainer.addCriteria(searchCriteria);
+                        }
```
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/support/SavedHistory.java]: Can we add a logging statement if the `searchCriteria` was null?

File: features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/topo/SearchResult.java [hunk 19]
```diff
@@ -28,16 +28,33 @@
...
+
+@XmlRootElement(name = "search")
+@XmlAccessorType(value = XmlAccessType.FIELD)
 public class SearchResult {
 
-    private final String m_id;
-    private final String m_namespace;
-    private final String m_label;
-    private final String m_query;
+	@XmlElement
```
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/topo/SearchResult.java]: We should add `name="<name>"`.

File: features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/topo/SearchResult.java [hunk 20]
```diff
@@ -28,16 +28,33 @@
...
+    private String m_id;
+	@XmlElement
```
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/topo/SearchResult.java]: We should add `name="<name>"`.

File: features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/topo/SearchResult.java [hunk 21]
```diff
@@ -28,16 +28,33 @@
...
+    private String m_namespace;
+	@XmlElement
```
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/topo/SearchResult.java]: We should add `name="<name>"`.

File: features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/topo/SearchResult.java [hunk 22]
```diff
@@ -28,16 +28,33 @@
...
+    private String m_label;
+	@XmlElement
```
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/topo/SearchResult.java]: We should add `name="<name>"`.

File: features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/topo/SearchResult.java [hunk 23]
```diff
@@ -28,16 +28,33 @@
...
+    private String m_query;
+	@XmlElement
```
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/topo/SearchResult.java]: We should add `name="<name>"`.

File: features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/topo/SearchResult.java [hunk 24]
```diff
@@ -28,16 +28,33 @@
...
     private boolean m_collapsible = false;
+	@XmlElement
```
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/topo/SearchResult.java]: We should add `name="<name>"`.

File: features/topology-map/org.opennms.features.topology.api/src/test/java/org/opennms/features/topology/api/support/SavedHistoryTest.java [hunk 25]
```diff
@@ -70,8 +77,64 @@ public void testMarshall() {
...
 				Collections.singleton(vert1),
-				settings
+				settings,
+				Collections.emptyList()
 		);
 		JAXB.marshal(savedHistory, System.out);
 	}
+
+	@Test
+	public void verifyCRCIncludesSearchResult() {
```
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.api/src/test/java/org/opennms/features/topology/api/support/SavedHistoryTest.java]: Can we add a comment to describe what the test is verifiing?

File: features/topology-map/org.opennms.features.topology.app/src/test/java/org/opennms/features/topology/app/internal/IpLikeSearchProviderTest.java [hunk 26]
```diff
@@ -0,0 +1,66 @@
...
+
+import com.google.common.collect.Lists;
+
+public class IpLikeSearchProviderTest {
+
+	private static final String namespace = "iplike";
+
+	@Test
+	public void verifyCriteriaFromQueryGeneration()
+	{
```
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.app/src/test/java/org/opennms/features/topology/app/internal/IpLikeSearchProviderTest.java]: Bracket should be at the end of the line before (-:

File: features/topology-map/plugins/org.opennms.features.topology.plugins.ncs/src/main/java/org/opennms/features/topology/plugins/ncs/NCSSearchProvider.java [hunk 27]
```diff
@@ -51,7 +52,7 @@
 import org.opennms.netmgt.model.ncs.NCSComponent;
 import org.opennms.netmgt.model.ncs.NCSComponentRepository;
 
-public class NCSSearchProvider extends AbstractSearchProvider implements SearchProvider {
+public class NCSSearchProvider extends AbstractSearchProvider implements SearchProvider, HistoryAwareSearchProvider {
```
Comment (mvrueden) [inline features/topology-map/plugins/org.opennms.features.topology.plugins.ncs/src/main/java/org/opennms/features/topology/plugins/ncs/NCSSearchProvider.java]: Should only implement HistoryAwareSearchProvider.

File: features/topology-map/plugins/org.opennms.features.topology.plugins.topo.history/src/test/java/org/opennms/features/topology/plugins/topo/BundleContextHistoryManagerTest.java [hunk 28]
```diff
@@ -201,12 +241,83 @@ public void testHistoryManager() throws IOException {
...
+    }
+
     private static Properties loadProperties() throws IOException {
         Properties props = new Properties();
         props.load(new FileInputStream(new File(DATA_FILE_NAME)));
         return props;
     }
 
+    private void initAvailableProvidersAndCriteria() {
+        CategoryProvider vertexProvider = new CategoryProvider() {
```
Comment (mvrueden) [inline features/topology-map/plugins/org.opennms.features.topology.plugins.topo.history/src/test/java/org/opennms/features/topology/plugins/topo/BundleContextHistoryManagerTest.java]: This looks fine to me, but why do we have a `CategoryProviderTest` in addition to this? We should remove the test/dummy implementation. This is true for all test provideres if they are not used anywhere else.

File: features/topology-map/plugins/org.opennms.features.topology.plugins.topo.history/src/test/java/org/opennms/features/topology/plugins/topo/BundleContextHistoryManagerTest.java [hunk 29]
```diff
@@ -258,10 +369,15 @@ private void setBehaviour(GraphContainer graphContainerMock) {
         		.andReturn(selectionManagerMock)
         		.anyTimes();
 
+        Criteria[] criteria = new Criteria[availableCriteria.size()];
+
         EasyMock
                 .expect(graphContainerMock.getCriteria())
-                .andReturn(new Criteria[0])
+                .andReturn(availableCriteria.toArray(criteria))
```
Comment (mvrueden) [inline features/topology-map/plugins/org.opennms.features.topology.plugins.topo.history/src/test/java/org/opennms/features/topology/plugins/topo/BundleContextHistoryManagerTest.java]: I think we should actually return `capturedCriteria.getValues()` here.

**Review by mvrueden — CHANGES_REQUESTED** [review]
This looks very good now 👍 
I found some more minor issues. 
Please address them. 
I will give this a spin and verify that it actually still works and also that the history looks good

**Comment by mvrueden** [comment]
I created a Pull Request to make the "login" and "logout" methods available to all tests. See https://github.com/OpenNMS/smoke-tests/pull/2 . We have to wait until it gets merged, though.

File: features/topology-map/plugins/org.opennms.features.topology.plugins.topo.history/src/test/java/org/opennms/features/topology/plugins/topo/BundleContextHistoryManagerTest.java (hunk 29 above)
Comment (pdzaccone) [reply features/topology-map/plugins/org.opennms.features.topology.plugins.topo.history/src/test/java/org/opennms/features/topology/plugins/topo/BundleContextHistoryManagerTest.java]: `capturedCriteria` object was introduced to intercept parameters from the method `addCriteria`. Objects from the `capturedCriteria` are later used in assertion.
If the method `getCriteria` returns the same values, then the comparison will always return true. The very idea of test will then be lost.
Besides, at the beginning the `capturedCriteria` is empty and the `getCriteria` method is called earlier than `addCriteria`

Comment (mvrueden) [reply features/topology-map/plugins/org.opennms.features.topology.plugins.topo.history/src/test/java/org/opennms/features/topology/plugins/topo/BundleContextHistoryManagerTest.java]: Then let me rephrase. In order that the test actually makes sense, we should mock the `addCriteria()`, `removeCriteria()` and `getCriteria()` methods, rather than always return a fixed list. Otherwise the whole purpose of the test is questionable.

File: features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/support/SavedHistory.java [hunk 30]
```diff
@@ -201,6 +227,19 @@ public void apply(GraphContainer graphContainer, Collection<HistoryOperation> op
                 LOG.warn("Failed to perform applyHistory() operation", e);
             }
         }
+
```
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/support/SavedHistory.java:206]: Can we add a description here on what we are doing and why?

File: features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/topo/SearchCriteria.java [hunk 31]
```diff
@@ -0,0 +1,33 @@
...
+ *
+ * For more information contact:
+ *     OpenNMS(R) Licensing <license@opennms.org>
+ *     http://www.opennms.org/
+ *     http://www.opennms.com/
+ *******************************************************************************/
+
+package org.opennms.features.topology.api.topo;
+
+public interface SearchCriteria extends CollapsibleCriteria {
```
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/topo/SearchCriteria.java]: Can we add a description here to help developers understand what the `SearchCriteria` is supposed to do?

File: smoke-test/src/test/java/org/opennms/smoketest/TopologyIT.java [hunk 32]
```diff
@@ -693,6 +694,37 @@ public void verifyPingOperation() throws InterruptedException, IOException {
...
+        logout();
+
+        login();
+        topologyUiPage.open();
+        focusedVertices = topologyUiPage.getFocusedVertices();
+        Assert.assertNotNull(focusedVertices);
+        Assert.assertEquals(1, focusedVertices.size());
+    }
+
+    private void login() {
```
Comment (mvrueden) [inline smoke-test/src/test/java/org/opennms/smoketest/TopologyIT.java]: This can be removed now

File: smoke-test/src/test/java/org/opennms/smoketest/TopologyIT.java [hunk 33]
```diff
@@ -693,6 +694,37 @@ public void verifyPingOperation() throws InterruptedException, IOException {
...
+        topologyUiPage.testCase.m_driver.get(topologyUiPage.testCase.getBaseUrl() + "opennms/login.jsp");
+        topologyUiPage.testCase.wait.until(ExpectedConditions.visibilityOfElementLocated(By.name("j_username")));
+        topologyUiPage.testCase.wait.until(ExpectedConditions.visibilityOfElementLocated(By.name("j_password")));
+        topologyUiPage.testCase.wait.until(ExpectedConditions.elementToBeClickable(By.name("Login")));
+        topologyUiPage.testCase.enterText(By.name("j_username"), new CharSequence[]{"admin"});
+        topologyUiPage.testCase.enterText(By.name("j_password"), new CharSequence[]{"admin"});
+        topologyUiPage.testCase.findElementByName("Login").click();
+    }
+
+    private void logout() {
```
Comment (mvrueden) [inline smoke-test/src/test/java/org/opennms/smoketest/TopologyIT.java]: This can be removed now

**Review by mvrueden — APPROVED** [review]

**Comment by mvrueden** [comment]
Should be merged to opennms/foundation-2017. Closing.
